deploy/containers/credhub_setup/auth.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+// This file contains code to deal with UAA / authentication.
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+func makeHTTPClientWithCA(serverName string, caCert []byte) (*http.Client, error) {
+	certPool, err := x509.SystemCertPool()
```

Comment:
SystemCertPool is a hidden dependency. Ideally, `makeHTTPClientWithCA` will be a public method of the `HttpClient` struct. You can have something like:

```go
type HttpClient struct {
  systemCertPool func() (*x509.CertPool, error)
}

func NewHttpClient(systemCertPool func() (*x509.CertPool, error)) *HttpClient {
  return &HttpClient{
    systemCertPool: systemCertPool,
  }
}

func (httpClient *HttpClient) MakeWithCACert(serverName string, caCert []byte) (*http.Client, error) {
  certPool, err := httpClient.systemCertPool()
  ...
}
```

When using it, just construct with:

```go
httpClient := uaa.NewHttpClient(x509.SystemCertPool)
```

And there is another option would be to have another constructor for `HttpClient`:

```go
func NewDefaultHttpClient() *HttpClient {
  return &HttpClient{
    systemCertPool: x509.SystemCertPool,
  }
}
```

This is very important in order to be able to unit test this and ensure long-term maintainability.

Reply:
Please, create a new issue to add unit tests to all this so we can cover 100%.